Source/OxyPlot/Series/ContourSeries.cs

```diff
@@ -486,100 +459,118 @@ private void AddContourLabels(Contour contour, ScreenPoint[] pts, ICollection<Co
-            var contourPoints = new List<DataPoint>();
-            int contourPointsCount = 0;
 
-            ContourSegment firstSegment = null;
-            int segmentCount = this.segments.Count;
-            while (segmentCount > 0)
+            static IEnumerable<SegmentPoint> GetPoints(ContourSegment segment)
             {
-                ContourSegment segment1 = null, segment2 = null;
+                var p1 = new SegmentPoint(segment.StartPoint);
+                var p2 = new SegmentPoint(segment.EndPoint);
+                p1.Partner = p2;
+                p2.Partner = p1;
+                yield return p1;
+                yield return p2;
+            }
 
-                if (firstSegment != null)
+            foreach (var group in this.segments.GroupBy(p => p.ContourLevel))
+            {
+                var level = group.Key;
+                var points = group.SelectMany(GetPoints).OrderBy(p => p.Point.X).ToList();
+
+                // first, go through the sorted points, find identical points and join them together 
+                for (var i = 0; i < points.Count - 1; i++)
                 {
-                    bool reverse;
+                    var currentPoint = points[i];
+                    if (currentPoint.Join != null)
+                    {
+                        continue;
+                    }
 
-                    // Find a segment that is connected to the head of the contour
-                    segment1 = this.FindConnectedSegment(contourPoints[0], firstSegment.ContourLevel, eps, out reverse);
-                    if (segment1 != null)
+                    var maxX = currentPoint.Point.X + eps;
+                    var i2 = 1;
+                    SegmentPoint joinPoint;
+
+                    // search for a point with the same coordinates (within eps) as the current point
+                    // as points are sorted by X, we typically only need to check the point immediately following the current point
+                    do
                     {
-                        contourPoints.Insert(0, reverse ? segment1.StartPoint : segment1.EndPoint);
-                        contourPointsCount++;
-                        this.segments.Remove(segment1);
-                        segmentCount--;
+                        if (i + i2 >= points.Count)
+                        {
+                            joinPoint = null;
+                            break;
+                        }
+
+                        joinPoint = points[i + i2];
+                        i2++;
+                        if (joinPoint.Join != null)
+                        {
+                            continue;
+                        }
+
+                        if (joinPoint.Point.X > maxX)
+                        {
+                            joinPoint = null;
+                            break;
+                        }
                     }
+                    while (Math.Abs(joinPoint.Point.Y - currentPoint.Point.Y) > eps);
```

Comment:
Any reason this isn't in the `if` at the top of the loop? I'm not sure I get why this is `>` rather than `<`.

Reply:
Well we go through the next couple of points after `currentPoint` _until_ we find one whose Y coordinate is within `eps` of the `currentPoint`s Y coordinate. That's what the `while` condition is checking for. Sure, this could be pulled into an `if` in the loop body, but I found it more concise that way.

Reply:
Ah, yeah, that makes sense.

Reply:
What may be confusing is the fact that usually this loop will only run exactly once, because in most cases the point immediately following `currentPoint` will be the matching one (because they are already sorted by X).

Reply:
I'd just completely missed that the work was done outside the loop. It's a bit confusing with all the exit conditions, but I can't think how to make it much clearer.